### PR TITLE
Transforms: Add join by fields

### LINKED ISSUE #56424: (transformations) Joining two tables doesn't allow you to select both right and left side fields
**What happened**:

I want to join two tables (two different queries) together with transformations. I tried to use the "join by field" but that only works if both tables have the field with the same name and type.

**What you expected to happen**:

I want to be able to select both the left and right side fields of the join. e.g. `WHERE a.field = b.other_field`.

**How to reproduce it (as minimally and precisely as possible)**:

* Get some table data (i was using github datasource, for tag data)
* Get some other table data (i was using github datasource, for release data)
* try to join those tables together with the transformations

**Anything else we need to know?**:

**Environment**:
- Grafana version: 9.x
- Data source type & version: any with tabular data
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by LudoVio**
**What is this feature?**

Add the possibility to select the field name for each dataFrame during a JoinByField tranform

**Why do we need this feature?**

Sometimes user want to join table based on columns that have different name. This is not possible right now.

**Who is this feature for?**

Everyone using tranform

**Which issue(s) does this PR fix?**:

Fixes #56424

### CODE REVIEW

File: packages/grafana-data/src/transformations/transformers/joinByField.test.ts [hunk 1]
```diff
@@ -1011,21 +1056,26 @@ describe('JOIN Transformer', () => {
...
       const frame1 = toDataFrame({
+        refId: 'frame1',
         fields: [
           { name: 'time', type: FieldType.time, values: [1] },
           { name: 'temperature', type: FieldType.number, values: [10] },
         ],
       });
 
       const frame2 = toDataFrame({
+        refId: 'frame2',
```
Comment (nmarrs) [inline packages/grafana-data/src/transformations/transformers/joinByField.test.ts:1078]: Great job on this! Would it be possible to include a gdev dashboard for this transform? We can deploy / share gdev dashboards via play.grafana.org. We can provide a link to the public dashboard in release notes so users can try out changes easily. They can also help reviewers of the PR test out the changes

Comment (nmarrs) [reply packages/grafana-data/src/transformations/transformers/joinByField.test.ts:1078]: see [this doc](https://docs.google.com/document/d/1Zt_MHSWCd9vNX84vdVcHi4y-WuoZB86tB8ezZl_i3Qs/edit?usp=sharing) I wrote up explaining more on gdev dashboards 👍

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts [hunk 2]
```diff
@@ -47,6 +47,11 @@ export interface JoinOptions {
    */
   joinBy?: FieldMatcher;
 
+  /**
+   * The fields to join on
+   */
+  fields?: { [key: string]: string };
```
Comment (ryantxu) [inline packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:53]: is this parallel to the "joinBy"?

Can the frame/field selection be part of a FieldMatcher (and config?)?  -- then we can use the logic/ui all the places that need to pick fields out of a `[]frame`

File: public/app/features/transformers/editors/JoinByFieldTransformerEditor.tsx [hunk 3]
```diff
@@ -1,31 +1,60 @@
...
+
+      if (options.fields) {
+        if (dataFrame.refId in options.fields) {
+          if (Object.keys(options.fields).length === 1) {
+            return;
+          }
+
+          delete options.fields[dataFrame.refId];
+        } else {
+          options.fields[daoptions.fieldstaFrame.refId] = dataFrame.fields[0].name;
```
Comment (itsmylife) [inline public/app/features/transformers/editors/JoinByFieldTransformerEditor.tsx:41]:
```suggestion
          options.fields[dataFrame.refId] = dataFrame.fields[0].name;
```

Comment (itsmylife) [reply public/app/features/transformers/editors/JoinByFieldTransformerEditor.tsx:41]: I think there was a mistake and this is the solution.

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts (hunk 2 above)
Comment (LudoVio) [reply packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:53]: Yes it is parallel to joinBy, which takes a FieldMatcher

Currently the transformer create FieldMatchers based on frame.refId and field.name. Do you suggest to allow passing FieldMatchers like for joinBy?

File: packages/grafana-data/src/transformations/transformers/joinByField.test.ts (hunk 1 above)
Comment (LudoVio) [reply packages/grafana-data/src/transformations/transformers/joinByField.test.ts:1078]: Added a dashboard

Screenshot_20230118_011125

**Review by itsmylife — APPROVED** [review]

**Comment by torkelo** [comment]
@LudoVio corrected milestone on this issue as there is no backports for this it cannot be in 9.3.4, 

also sadly have to revert this as it needs some more work, https://github.com/grafana/grafana/pull/62278

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts (hunk 2 above)
Comment (torkelo) [reply packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:53]: think this needs a bit of rethink

**Pull request merged**